File: android/src/main/java/ai/annadata/plugin/capacitor/LlamaCppPlugin.java
package ai.annadata.plugin.capacitor;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.JSArray;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.Map;
import org.json.JSONException;

@CapacitorPlugin(name = "LlamaCpp")
public class LlamaCppPlugin extends Plugin {
    private static final String TAG = "LlamaCppPlugin";

    private LlamaCpp implementation = new LlamaCpp();

    @Override
    public void load() {
        super.load();
        Log.i(TAG, "LlamaCppPlugin loaded successfully");
    }

    // MARK: - Core initialization and management

    @PluginMethod
    public void toggleNativeLog(PluginCall call) {
        boolean enabled = call.getBoolean("enabled", false);
        implementation.toggleNativeLog(enabled, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void setContextLimit(PluginCall call) {
        int limit = call.getInt("limit", 10);
        implementation.setContextLimit(limit, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void modelInfo(PluginCall call) {
        String path = call.getString("path", "");
        JSArray skipArray = call.getArray("skip");
        String[] skip = new String[0];
        if (skipArray != null) {
            skip = new String[skipArray.length()];
            for (int i = 0; i < skipArray.length(); i++) {
                try {
                    skip[i] = skipArray.getString(i);
                } catch (JSONException e) {
                    skip[i] = "";
                }
            }
        }

        implementation.modelInfo(path, skip, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void initContext(PluginCall call) {
        Log.i(TAG, "initContext called with contextId: " + call.getInt("contextId", 0));
        int contextId = call.getInt("contextId", 0);
        JSObject params = call.getObject("params", new JSObject());

        implementation.initContext(contextId, params, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void releaseContext(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.releaseContext(contextId, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void releaseAllContexts(PluginCall call) {
        implementation.releaseAllContexts(result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Chat and completion

    @PluginMethod
    public void getFormattedChat(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String messages = call.getString("messages", "");
        String chatTemplate = call.getString("chatTemplate");
        JSObject params = call.getObject("params");

        implementation.getFormattedChat(contextId, messages, chatTemplate, params, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void completion(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        JSObject params = call.getObject("params", new JSObject());

        implementation.completion(contextId, params, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void stopCompletion(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.stopCompletion(contextId, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Session management

    @PluginMethod
    public void loadSession(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String path = call.getString("path", "");

        implementation.loadSession(contextId, path, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void saveSession(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String path = call.getString("path", "");
        int size = call.getInt("size", -1);

        implementation.saveSession(contextId, path, size, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("tokensSaved", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Tokenization

    @PluginMethod
    public void tokenize(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String text = call.getString("text", "");
        JSArray imagePathsArray = call.getArray("imagePaths");
        String[] imagePaths = new String[0];
        if (imagePathsArray != null) {
            imagePaths = new String[imagePathsArray.length()];
            for (int i = 0; i < imagePathsArray.length(); i++) {
                try {
                    imagePaths[i] = imagePathsArray.getString(i);
                } catch (JSONException e) {
                    imagePaths[i] = "";
                }
            }
        }

        implementation.tokenize(contextId, text, imagePaths, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void detokenize(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        JSArray tokensArray = call.getArray("tokens");
        Integer[] tokens = new Integer[0];
        if (tokensArray != null) {
            tokens = new Integer[tokensArray.length()];
            for (int i = 0; i < tokensArray.length(); i++) {
                try {
                    tokens[i] = tokensArray.getInt(i);
                } catch (JSONException e) {
                    tokens[i] = 0;
                }
            }
        }

        implementation.detokenize(contextId, tokens, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("text", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Embeddings and reranking

    @PluginMethod
    public void embedding(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String text = call.getString("text", "");
        JSObject params = call.getObject("params", new JSObject());

        implementation.embedding(contextId, text, params, result -> {
            if (result.isSuccess()) {
                JSObject jsResult = new JSObject();
                Map<String, Object> data = result.getData();
                for (Map.Entry<String, Object> entry : data.entrySet()) {
                    jsResult.put(entry.getKey(), entry.getValue());
                }
                call.resolve(jsResult);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void rerank(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String query = call.getString("query", "");
        JSArray documentsArray = call.getArray("documents");
        String[] documents = new String[0];
        if (documentsArray != null) {
            documents = new String[documentsArray.length()];
            for (int i = 0; i < documentsArray.length(); i++) {
                try {
                    documents[i] = documentsArray.getString(i);
                } catch (JSONException e) {
                    documents[i] = "";
                }
            }
        }
        JSObject params = call.getObject("params");

        implementation.rerank(contextId, query, documents, params, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("results", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Benchmarking

    @PluginMethod
    public void bench(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        int pp = call.getInt("pp", 0);
        int tg = call.getInt("tg", 0);
        int pl = call.getInt("pl", 0);
        int nr = call.getInt("nr", 0);

        implementation.bench(contextId, pp, tg, pl, nr, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("result", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - LoRA adapters

    @PluginMethod
    public void applyLoraAdapters(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        JSArray loraAdaptersArray = call.getArray("loraAdapters");
        JSObject[] loraAdapters = new JSObject[0];
        if (loraAdaptersArray != null) {
            loraAdapters = new JSObject[loraAdaptersArray.length()];
            for (int i = 0; i < loraAdaptersArray.length(); i++) {
                // For now, create empty JSObjects since the exact method is unclear
                loraAdapters[i] = new JSObject();
            }
        }

        implementation.applyLoraAdapters(contextId, loraAdapters, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void removeLoraAdapters(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.removeLoraAdapters(contextId, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void getLoadedLoraAdapters(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.getLoadedLoraAdapters(contextId, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("adapters", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Multimodal methods

    @PluginMethod
    public void initMultimodal(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        JSObject params = call.getObject("params", new JSObject());
        String path = params.getString("path", "");
        boolean useGpu = params.getBoolean("use_gpu", true);

        implementation.initMultimodal(contextId, path, useGpu, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("success", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void isMultimodalEnabled(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.isMultimodalEnabled(contextId, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("enabled", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void getMultimodalSupport(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.getMultimodalSupport(contextId, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("support", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void releaseMultimodal(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.releaseMultimodal(contextId, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - TTS methods

    @PluginMethod
    public void initVocoder(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        JSObject params = call.getObject("params", new JSObject());
        String path = params.getString("path", "");
        Integer nBatch = params.getInteger("n_batch");

        implementation.initVocoder(contextId, path, nBatch, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("success", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void isVocoderEnabled(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.isVocoderEnabled(contextId, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("enabled", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void getFormattedAudioCompletion(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String speakerJsonStr = call.getString("speakerJsonStr", "");
        String textToSpeak = call.getString("textToSpeak", "");

        implementation.getFormattedAudioCompletion(contextId, speakerJsonStr, textToSpeak, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("completion", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void getAudioCompletionGuideTokens(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        String textToSpeak = call.getString("textToSpeak", "");

        implementation.getAudioCompletionGuideTokens(contextId, textToSpeak, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("tokens", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void decodeAudioTokens(PluginCall call) {
        int contextId = call.getInt("contextId", 0);
        JSArray tokensArray = call.getArray("tokens");
        Integer[] tokens = new Integer[0];
        if (tokensArray != null) {
            tokens = new Integer[tokensArray.length()];
            for (int i = 0; i < tokensArray.length(); i++) {
                try {
                    tokens[i] = tokensArray.getInt(i);
                } catch (JSONException e) {
                    tokens[i] = 0;
                }
            }
        }

        implementation.decodeAudioTokens(contextId, tokens, result -> {
            if (result.isSuccess()) {
                JSObject ret = new JSObject();
                ret.put("decodedTokens", result.getData());
                call.resolve(ret);
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    @PluginMethod
    public void releaseVocoder(PluginCall call) {
        int contextId = call.getInt("contextId", 0);

        implementation.releaseVocoder(contextId, result -> {
            if (result.isSuccess()) {
                call.resolve();
            } else {
                call.reject(result.getError().getMessage());
            }
        });
    }

    // MARK: - Events

    @PluginMethod
    public void addListener(PluginCall call) {
        String eventName = call.getString("eventName", "");
        // Note: In Capacitor, event listeners are typically handled differently
        // This is a placeholder for the event system
        call.resolve();
    }

    @PluginMethod
    public void removeAllListeners(PluginCall call) {
        String eventName = call.getString("eventName", "");
        // Note: In Capacitor, event listeners are typically handled differently
        // This is a placeholder for the event system
        call.resolve();
    }
}

File: ios/Sources/LlamaCppPlugin/LlamaCppPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LlamaCppPlugin)
public class LlamaCppPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LlamaCppPlugin"
    public let jsName = "LlamaCpp"
    public let pluginMethods: [CAPPluginMethod] = [
        // Core initialization and management
        CAPPluginMethod(name: "toggleNativeLog", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setContextLimit", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "modelInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initContext", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseContext", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseAllContexts", returnType: CAPPluginReturnPromise),
        
        // Chat and completion
        CAPPluginMethod(name: "getFormattedChat", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "completion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopCompletion", returnType: CAPPluginReturnPromise),
        
        // Session management
        CAPPluginMethod(name: "loadSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveSession", returnType: CAPPluginReturnPromise),
        
        // Tokenization
        CAPPluginMethod(name: "tokenize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detokenize", returnType: CAPPluginReturnPromise),
        
        // Embeddings and reranking
        CAPPluginMethod(name: "embedding", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rerank", returnType: CAPPluginReturnPromise),
        
        // Benchmarking
        CAPPluginMethod(name: "bench", returnType: CAPPluginReturnPromise),
        
        // LoRA adapters
        CAPPluginMethod(name: "applyLoraAdapters", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeLoraAdapters", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLoadedLoraAdapters", returnType: CAPPluginReturnPromise),
        
        // Multimodal methods
        CAPPluginMethod(name: "initMultimodal", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isMultimodalEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getMultimodalSupport", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseMultimodal", returnType: CAPPluginReturnPromise),
        
        // TTS methods
        CAPPluginMethod(name: "initVocoder", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isVocoderEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getFormattedAudioCompletion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAudioCompletionGuideTokens", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "decodeAudioTokens", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseVocoder", returnType: CAPPluginReturnPromise),
        
        // Events
        CAPPluginMethod(name: "addListener", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeAllListeners", returnType: CAPPluginReturnPromise)
    ]
    
    private let implementation = LlamaCpp()

    // MARK: - Core initialization and management
    
    @objc func toggleNativeLog(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? false
        implementation.toggleNativeLog(enabled: enabled) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func setContextLimit(_ call: CAPPluginCall) {
        let limit = call.getInt("limit") ?? 0
        implementation.setContextLimit(limit: limit) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func modelInfo(_ call: CAPPluginCall) {
        let path = call.getString("path") ?? ""
        let skip = call.getArray("skip", String.self) ?? []
        
        implementation.modelInfo(path: path, skip: skip) { result in
            switch result {
            case .success(let modelInfo):
                call.resolve(modelInfo)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func initContext(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let params = call.getObject("params") ?? [:]
        
        implementation.initContext(contextId: contextId, params: params) { result in
            switch result {
            case .success(let context):
                call.resolve(context)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func releaseContext(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.releaseContext(contextId: contextId) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func releaseAllContexts(_ call: CAPPluginCall) {
        implementation.releaseAllContexts { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Chat and completion
    
    @objc func getFormattedChat(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let messages = call.getString("messages") ?? ""
        let chatTemplate = call.getString("chatTemplate")
        let params = call.getObject("params")
        
        implementation.getFormattedChat(
            contextId: contextId,
            messages: messages,
            chatTemplate: chatTemplate,
            params: params
        ) { result in
            switch result {
            case .success(let formattedChat):
                call.resolve(formattedChat)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func completion(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let params = call.getObject("params") ?? [:]
        
        implementation.completion(contextId: contextId, params: params) { result in
            switch result {
            case .success(let completionResult):
                call.resolve(completionResult)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func stopCompletion(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.stopCompletion(contextId: contextId) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Session management
    
    @objc func loadSession(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let filepath = call.getString("filepath") ?? ""
        
        implementation.loadSession(contextId: contextId, filepath: filepath) { result in
            switch result {
            case .success(let sessionResult):
                call.resolve(sessionResult)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func saveSession(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let filepath = call.getString("filepath") ?? ""
        let size = call.getInt("size") ?? -1
        
        implementation.saveSession(contextId: contextId, filepath: filepath, size: size) { result in
            switch result {
            case .success(let tokensSaved):
                call.resolve(["tokensSaved": tokensSaved])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Tokenization
    
    @objc func tokenize(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let text = call.getString("text") ?? ""
        let imagePaths = call.getArray("imagePaths", String.self) ?? []
        
        implementation.tokenize(contextId: contextId, text: text, imagePaths: imagePaths) { result in
            switch result {
            case .success(let tokenizeResult):
                call.resolve(tokenizeResult)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func detokenize(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let tokens = call.getArray("tokens", Int.self) ?? []
        
        implementation.detokenize(contextId: contextId, tokens: tokens) { result in
            switch result {
            case .success(let text):
                call.resolve(["text": text])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Embeddings and reranking
    
    @objc func embedding(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let text = call.getString("text") ?? ""
        let params = call.getObject("params") ?? [:]
        
        implementation.embedding(contextId: contextId, text: text, params: params) { result in
            switch result {
            case .success(let embeddingResult):
                call.resolve(embeddingResult)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func rerank(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let query = call.getString("query") ?? ""
        let documents = call.getArray("documents", String.self) ?? []
        let params = call.getObject("params")
        
        implementation.rerank(contextId: contextId, query: query, documents: documents, params: params) { result in
            switch result {
            case .success(let rerankResults):
                call.resolve(["results": rerankResults])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Benchmarking
    
    @objc func bench(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let pp = call.getInt("pp") ?? 0
        let tg = call.getInt("tg") ?? 0
        let pl = call.getInt("pl") ?? 0
        let nr = call.getInt("nr") ?? 0
        
        implementation.bench(contextId: contextId, pp: pp, tg: tg, pl: pl, nr: nr) { result in
            switch result {
            case .success(let benchResult):
                call.resolve(["result": benchResult])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - LoRA adapters
    
    @objc func applyLoraAdapters(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let loraAdapters = call.getArray("loraAdapters", [String: Any].self) ?? []
        
        implementation.applyLoraAdapters(contextId: contextId, loraAdapters: loraAdapters) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func removeLoraAdapters(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.removeLoraAdapters(contextId: contextId) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func getLoadedLoraAdapters(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.getLoadedLoraAdapters(contextId: contextId) { result in
            switch result {
            case .success(let adapters):
                call.resolve(["adapters": adapters])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Multimodal methods
    
    @objc func initMultimodal(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let params = call.getObject("params") ?? [:]
        let path = params["path"] as? String ?? ""
        let useGpu = params["use_gpu"] as? Bool ?? true
        
        implementation.initMultimodal(contextId: contextId, path: path, useGpu: useGpu) { result in
            switch result {
            case .success(let success):
                call.resolve(["success": success])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func isMultimodalEnabled(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.isMultimodalEnabled(contextId: contextId) { result in
            switch result {
            case .success(let enabled):
                call.resolve(["enabled": enabled])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func getMultimodalSupport(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.getMultimodalSupport(contextId: contextId) { result in
            switch result {
            case .success(let support):
                call.resolve(support)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func releaseMultimodal(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.releaseMultimodal(contextId: contextId) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - TTS methods
    
    @objc func initVocoder(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let params = call.getObject("params") ?? [:]
        let path = params["path"] as? String ?? ""
        let nBatch = params["n_batch"] as? Int
        
        implementation.initVocoder(contextId: contextId, path: path, nBatch: nBatch) { result in
            switch result {
            case .success(let success):
                call.resolve(["success": success])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func isVocoderEnabled(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.isVocoderEnabled(contextId: contextId) { result in
            switch result {
            case .success(let enabled):
                call.resolve(["enabled": enabled])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func getFormattedAudioCompletion(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let speakerJsonStr = call.getString("speakerJsonStr") ?? ""
        let textToSpeak = call.getString("textToSpeak") ?? ""
        
        implementation.getFormattedAudioCompletion(
            contextId: contextId,
            speakerJsonStr: speakerJsonStr,
            textToSpeak: textToSpeak
        ) { result in
            switch result {
            case .success(let audioCompletion):
                call.resolve(audioCompletion)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func getAudioCompletionGuideTokens(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let textToSpeak = call.getString("textToSpeak") ?? ""
        
        implementation.getAudioCompletionGuideTokens(contextId: contextId, textToSpeak: textToSpeak) { result in
            switch result {
            case .success(let tokens):
                call.resolve(["tokens": tokens])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func decodeAudioTokens(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        let tokens = call.getArray("tokens", Int.self) ?? []
        
        implementation.decodeAudioTokens(contextId: contextId, tokens: tokens) { result in
            switch result {
            case .success(let decodedTokens):
                call.resolve(["decodedTokens": decodedTokens])
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func releaseVocoder(_ call: CAPPluginCall) {
        let contextId = call.getInt("contextId") ?? 0
        
        implementation.releaseVocoder(contextId: contextId) { result in
            switch result {
            case .success:
                call.resolve()
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Events
    
    @objc func addListener(_ call: CAPPluginCall) {
        let eventName = call.getString("eventName") ?? ""
        // Note: In Capacitor, event listeners are typically handled differently
        // This is a placeholder for the event system
        call.resolve()
    }
    
    @objc func removeAllListeners(_ call: CAPPluginCall) {
        let eventName = call.getString("eventName") ?? ""
        // Note: In Capacitor, event listeners are typically handled differently
        // This is a placeholder for the event system
        call.resolve()
    }
}
